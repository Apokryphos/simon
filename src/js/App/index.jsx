import React from 'react';
import PropTypes from 'prop-types';
import Title from '../Components/Title';
import RadialList from '../Components/RadialList';
import ResetButton from '../Components/ResetButton';
import SequenceCount from '../Components/SequenceCount';
import StrictButton from '../Components/StrictButton';
import Victory from '../Components/Victory';
import Game from '../Game';
import GameController from '../GameController';

class App extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      controller: new GameController(new Game(8, props.random)),
      sequenceLength: 1,
      strict: false,
      victory: false,
    };

    this.handleButtonClick = this.handleButtonClick.bind(this);
    this.toggleStrict = this.toggleStrict.bind(this);
    this.reset = this.reset.bind(this);
  }

  componentDidMount() {
    this.state.controller.playSequence();
  }

  handleButtonClick(buttonIndex) {
    this.state.controller.submit(buttonIndex);
    this.setState(prevState => ({
      sequenceLength: prevState.controller.getSequenceLength(),
      victory: prevState.controller.getVictory(),
    }));
  }

  reset() {
    this.setState((prevState) => {
      prevState.controller.reset();
      prevState.controller.playSequence();
      return {
        sequenceLength: prevState.controller.getSequenceLength(),
        victory: false,
      };
    });
  }

  toggleStrict() {
    this.setState((prevState) => {
      prevState.controller.toggleStrict();
      return {
        strict: prevState.controller.getStrict(),
      };
    });
  }

  render() {
    let display = null;
    if (this.state.victory) {
      display = <Victory />;
    } else {
      display = <SequenceCount count={this.state.sequenceLength} />;
    }

    return (
      <div id="app">
        <RadialList gameController={this.state.controller} onClick={this.handleButtonClick} />
        <div className="menu-ui-container">
          <Title text="Simon" />
          {display}
          <ResetButton onClick={this.reset} />
          <StrictButton enabled={this.state.strict} onClick={this.toggleStrict} />
        </div>
      </div>
    );
  }
}

App.propTypes = {
  random: PropTypes.func,
};

App.defaultProps = {
  random: null,
};

export default App;
