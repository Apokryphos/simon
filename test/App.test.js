import React from 'react';
import { shallow } from 'enzyme';
import App from 'App';
import { getMatchButton } from './test-util';

describe('App rendering', () => {
  it('Renders OK', () => {
    const wrapper = shallow(<App random={() => 0} />);
    wrapper.instance().forceUpdate();
    wrapper.instance().render();
    expect(wrapper).toMatchSnapshot();
  });

  it('Renders OK after victory', () => {
    const wrapper = shallow(<App random={() => 0} />);

    const gameController = wrapper.state('controller');

    expect(gameController.getVictory() === false);
    wrapper.instance().render();

    for (let v = 0; v < gameController.getVictoryCount(); ++v) {
      expect(gameController.getSequenceLength === v);
      gameController.submit(getMatchButton(gameController.game));
    }

    wrapper.instance().render();
    expect(gameController.getVictory() === true);
    expect(wrapper).toMatchSnapshot();
  });

  it('handleButtonClick submits correctly', () => {
    const wrapper = shallow(<App />);

    expect(wrapper.state().sequenceLength === 1);

    const matchButton = wrapper.state('controller').game.getSequence()[0];
    wrapper.instance().handleButtonClick(matchButton);

    expect(wrapper.state().sequenceLength === 2);
  });

  it('reset method resets sequence length', (done) => {
    const wrapper = shallow(<App />);
    wrapper.instance().reset();

    const callback = () => {
      expect(wrapper.state().sequenceLength === wrapper.state().controller.getSequenceLength());
      done();
    };

    wrapper.setState(prevState => prevState, callback);
  });

  it('toggleStrict method', () => {
    const wrapper = shallow(<App />);
    const strict = wrapper.state('strict');
    wrapper.instance().toggleStrict();
    expect(wrapper.state().strict !== strict);
  });
});
