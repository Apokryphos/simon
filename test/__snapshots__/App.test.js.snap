// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App rendering Renders OK 1`] = `
<div
  id="app"
>
  <RadialList
    buttonCount={8}
    gameController={
      t {
        "_events": Object {
          "activateGameButton": Array [],
        },
        "game": Game {
          "buttonCount": 8,
          "getRandomButton": [Function],
          "sequence": Array [
            0,
          ],
          "sequenceIndex": 0,
          "strict": false,
        },
        "playDelay": 1100,
        "state": 0,
        "victoryCount": 20,
      }
    }
    onClick={[Function]}
  />
  <div
    className="menu-ui-container"
  >
    <Title
      text="Simon"
    />
    <SequenceCount
      count={1}
    />
    <ResetButton
      onClick={[Function]}
    />
    <StrictButton
      enabled={false}
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`App rendering Renders OK after victory 1`] = `
<div
  id="app"
>
  <RadialList
    buttonCount={8}
    gameController={
      t {
        "_events": Object {
          "activateGameButton": Array [],
        },
        "game": Game {
          "buttonCount": 8,
          "getRandomButton": [Function],
          "sequence": Array [
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          "sequenceIndex": 5,
          "strict": false,
        },
        "playDelay": 1100,
        "state": 0,
        "victoryCount": 20,
      }
    }
    onClick={[Function]}
  />
  <div
    className="menu-ui-container"
  >
    <Title
      text="Simon"
    />
    <SequenceCount
      count={1}
    />
    <ResetButton
      onClick={[Function]}
    />
    <StrictButton
      enabled={false}
      onClick={[Function]}
    />
  </div>
</div>
`;
